[FX] Refactor unique name handling

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48157 [WIP][FX] Revamp node unique name logic + add additional uniquing in codegen
* **#48205 [FX] Refactor unique name handling**

Differential Revision: [D25068934](https://our.internmc.facebook.com/intern/diff/D25068934)

Commit:
[FX] Refactor unique name handling

[ghstack-poisoned]

Commit:
Update on "[FX] Refactor unique name handling"


Differential Revision: [D25068934](https://our.internmc.facebook.com/intern/diff/D25068934)

[ghstack-poisoned]

zdevito commented on torch/fx/graph.py:
```diff
@@ -440,39 +426,40 @@ def output(self, result: Argument, type_expr: Optional[Any] = None):
+        return op
+
+    def _create_unique_name(self, candidate : str) -> str:
+        candidate = candidate.replace('.', '_')
```
nit: this is redundant with the regex below

zdevito commented on torch/fx/graph.py:
```diff
@@ -440,39 +426,40 @@ def output(self, result: Argument, type_expr: Optional[Any] = None):
-        i = self._used_names[op] = self._used_names[op] + 1
-        return f'{op}_{i}'
+        candidate = re.sub('[^0-9a-zA-Z_]+', '_', candidate)
+        candidate = _snake_case(candidate)
```
snake casing should probably go in _target_to_string. We want to take names as is when they are valid identifiers if they were provided as part of the `create_node` API.

zdevito commented on torch/fx/proxy.py:
```diff
@@ -154,7 +154,7 @@ def __torch_function__(self, orig_method, types, args=None, kwargs=None):
         else:
             return self.tracer.create_proxy('call_function', orig_method, args, kwargs,
-                                            name=self.tracer.graph._name(orig_method.__name__))
+                                            name=orig_method.__name__)
```
With the other suggestion to move the snake case stuff, you probably still want to call _target_to_str here so that you get a reasonable snake_case name.

zdevito reviewed (approved):
Looks good

Commit:
Update on "[FX] Refactor unique name handling"


Differential Revision: [D25068934](https://our.internmc.facebook.com/intern/diff/D25068934)

[ghstack-poisoned]

Commit:
Update on "[FX] Refactor unique name handling"


Differential Revision: [D25068934](https://our.internmc.facebook.com/intern/diff/D25068934)

[ghstack-poisoned]
